templates/.circleci/config.yml

```diff
@@ -0,0 +1,237 @@
+          root: /home/circleci/project
+          paths:
+            - build_<< parameters.config >>
+
+      - save_cache: # Store cache
+          key: v1-pkg-cache-{{ checksum "package-lock.json" }}
+          paths:
+            - "node_modules"
+
+      # Requires the SLACK_WEBHOOK env var to be set
+      # - slack/notify-on-failure
+
+  deploy:
+    docker:
+      - image: cimg/python:3.6
+    parameters:
+      config:
+        description: |
+          The config file you want to use to build your gatsby site.
+          E.g. 'development' would load configurations from '.env.development'
+
+          Note: .env.* file should not be added checked into source control rather should be
+          through your CircleCi's build environtemt variables see: https://circleci.com/docs/2.0/env-vars/#environment-variable-usage-options
+        type: enum
+        enum: ["development", "staging", "production"]
+        default: production
+      bucket:
+        description: The S3 Bucket to sync files to.
+        type: string
+      region:
+        type: string
+        default: ''
+        description: |
+          The region to use for AWS operations.
+    steps:
+      - attach_workspace:
+          at: *workspace
+      - run: # Required by aws-cli
+          name: Set AWS region var
+          command: echo 'export AWS_REGION=<< parameters.region >>' >> $BASH_ENV
+      - aws-s3/sync:
+          from: build_<< parameters.config >>
+          overwrite: true
+          to: 's3://<< parameters.bucket >>'
+      - run:
+          name: Invalidate Cloudfront
+          command: |
+            export DIST_ID=$(aws cloudfront list-distributions --query "DistributionList.Items[?Aliases.Items[?contains(@, '<< parameters.bucket >>')]].Id | [0]" | tr -d '"')
+            aws cloudfront create-invalidation --distribution-id ${DIST_ID} --paths "/*"
+
+workflows:
+    version: 2
+    # The main workflow. Check out the code, build it, push it, deploy to staging, test, deploy to production
+    build_test_and_deploy:
+      jobs:
+        - checkout_code
+
+        - unit_test:
+            requires:
+              - checkout_code
+
+        - build:
+            name: build_staging
+            config: staging
+            requires:
+              - checkout_code
+
+        - build:
+            name: build_production
+            config: production
+            requires:
+              - checkout_code
+
+        - deploy:
+            name: deploy_staging
+            config: staging
+            requires:
+              - build_staging
+            bucket: "<% index .Params `stagingHostRoot` %>"
+            region: "<% index .Params `region` %>"
+            # Depending on your development flow you may want to also only deploy to staging on master.
+            # filters:
+            #   branches:
+            #     only:  # only branches matching the below regex filters will run
+            #       - /^master$/
+
+        - deploy:
```

Comment:
There should probably be an approval step. If it deploys to staging and then immediately to production there's not much point in having the staging environment.